File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/solving/deduction/algorithm/HintEliminationAlgorithm.java
package com.github.ilikeyourhat.sudokusolver.solving.deduction.algorithm;

import com.github.ilikeyourhat.sudokusolver.model.Field;
import com.github.ilikeyourhat.sudokusolver.model.Region;
import com.github.ilikeyourhat.sudokusolver.model.hint.SudokuHintGrid;

import java.util.List;

public class HintEliminationAlgorithm extends DeductionAlgorithm {

    public static class Factory implements DeductionAlgorithm.Factory {

        public HintEliminationAlgorithm instance(List<Region> regions, SudokuHintGrid sudokuHintGrid) {
            return new HintEliminationAlgorithm(regions, sudokuHintGrid);
        }

    }

    private HintEliminationAlgorithm(List<Region> regions, SudokuHintGrid possibilities) {
        super(regions, possibilities);
    }

    @Override
    protected void solve(Region region) {
        for (Field field : region.fullFields()) {
            ensureFullFieldHaveNoHints(field);
            removeHintFromRegion(region, field.value());
        }
    }

    private void ensureFullFieldHaveNoHints(Field field) {
        if (!possibilities.isEmpty(field)) {
            possibilities.clear(field);
            notifyChange();
        }
    }

    private void removeHintFromRegion(Region region, int value) {
        for (Field field : region) {
            if (possibilities.remove(field, value)) {
                notifyChange();
            }
        }
    }

}

File: kudoku/src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/algorithm/HintEliminationAlgorithm.kt
package io.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import io.github.ilikeyourhat.kudoku.model.Field
import io.github.ilikeyourhat.kudoku.model.Region
import io.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid

class HintEliminationAlgorithm(
    regions: List<Region>,
    possibilities: SudokuHintGrid
) : DeductionAlgorithm(regions, possibilities) {

    class Factory : DeductionAlgorithm.Factory {
        override fun instance(regions: List<Region>, possibilities: SudokuHintGrid): HintEliminationAlgorithm {
            return HintEliminationAlgorithm(regions, possibilities)
        }
    }

    override fun solve(): Boolean {
        return regions
            .map { solve(it) }
            .any { it }
    }

    override fun solve(region: Region): Boolean {
        var changed = false
        for (field in region.fullFields()) {
            changed = changed or
                    ensureFullFieldHaveNoHints(field) or
                    removeHintFromRegion(region, field.value())
        }
        return changed
    }

    private fun ensureFullFieldHaveNoHints(field: Field): Boolean {
        if (!possibilities.isEmpty(field)) {
            possibilities.clear(field)
            return true
        }
        return false
    }

    private fun removeHintFromRegion(region: Region, value: Int): Boolean {
        var changed = false
        for (field in region) {
            changed = changed or possibilities.remove(field, value)
        }
        return changed
    }
}
